### PR TITLE
Invalid expressions for regexp parameters

### LINKED ISSUE #62: Regex groups seem to break it
Had a  regex with a group inside which breaks the parsing `/foo/:id(([a-f0-9]{32},?)+)`

``` js
var find = require('find-my-way')()

find.get('/foo/:id(([a-f0-9]{32},?)+)', function () {})
```

### LINKED ISSUE #63: Unclosed regex spins the cpu for ever
The following unclosed regex makes find-my-way spin the cpu forever. Throwing an exception could be nicer :)

``` js
var find = require('find-my-way')()

find.get('/foo/:id(a', function () {})
```

### PR DESCRIPTION
**Pull request by absolux**
Fixes #63 by throwing a `TypeError` error if the regexp expressions is wrong
Fixes #62

### CODE REVIEW

**Comment by absolux** [comment]
@delvedor I'm not using my laptop, so I push directly on Github.
I have updated the test case, and now working on the fix.

**Comment by absolux** [comment]
@delvedor What do you think about throwing errors for invalid paths ?

**Comment by mcollina** [comment]
> What do you think about throwing errors for invalid paths?

During insertion or at runtime?

**Comment by mcollina** [comment]
good for me.

File: index.js [hunk 1]
```diff
@@ -88,7 +88,9 @@ Router.prototype._on = function _on (method, path, handler, store) {
         isRegex = isRegex || path[i] === '('
         if (isRegex) {
           i = path.indexOf(')', i) + 1
-          break
+          if (i) break
+
+          throw new TypeError('Invalid parameter regexp expression')
```
Comment (delvedor) [inline index.js]: Can you also add the path to the error message?
It will be easier to fix.

Comment (absolux) [reply index.js]: Done

File: test/issue-63.test.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+const t = require('tap')
+const factory = require('../')
+
+t.test('issue-63', (t) => {
+  t.plan(1)
+
+  const fmw = factory()
+
+  t.throws(function () {
+    fmw.on('GET', '/foo/:id(a', function () {})
```
Comment (delvedor) [inline test/issue-63.test.js]: Can you also check the expected error message?
Something like
```js
try  {
  fmw.on('GET', '/foo/:id(a', function () {})
  t.fail('should fail')
} catch (err) {
  t.is(err.message, 'Invalid regexp expression in "/foo/:id(a"')
}
```

**Comment by delvedor** [comment]
@mafintosh does this fix both your issues?

File: index.js [hunk 3]
```diff
@@ -496,3 +496,34 @@ function getWildcardNode (node, method, path, len) {
...
+function getClosingParenthensePosition (path, idx) {
+  // `path.indexOf()` will always return the first position of the closing parenthese,
+  // but it's inefficient for grouped or wrong regexp expressions.
+  // see issues #62 and #63 for more info
+
+  var parentheses = 1
+
+  while (idx < path.length) {
+    idx++
+
```
Comment (mafintosh) [inline index.js]: if the next char is `\` then you need to ignore the next one for it to be spec compliant :)

```
if (path[idx] === '\\') {
  continue
}
```

Comment (absolux) [reply index.js]: Hi @mafintosh can give a use case using backslash. thanks

Comment (mafintosh) [reply index.js]: Sure, `find.get(/foo/(bar/\(baz\)))`, matches '/foo/bar(baz)'.

**Comment by mafintosh** [comment]
@delvedor ya and added a small comment

**Comment by mafintosh** [comment]
LTGM from otherwise

**Comment by absolux** [comment]
@delvedor it was a mistake, I reverted it
@mafintosh could you review the new changes

**Comment by delvedor** [comment]
Landed in `v1.10.4` :)

**Pull request merged**